RecoHGCal/TICL/plugins/TracksterLinkingbyGNN.cc

```diff
@@ -0,0 +1,443 @@
+                                           std::vector<std::vector<unsigned int>> &outputSuperclusters,
+                                           std::vector<std::vector<unsigned int>> &linkedTracksterIdToInputTracksterId)
+{
+    LogDebug("HGCalTICLSuperclustering") << "Linking Algo by " << nnVersion_ << std::endl;
+    // Create trackster graph
+    auto TICL_graph_result = produce_ticl_graph(input.tracksters, deltaEtaWindow_, deltaPhiWindow_);
+
+    // Access the TICLGraph object through the unique pointer
+    TICLGraph *ticlGraph = TICL_graph_result.get();
+
+    // Access the corresponding DNNVersion enum
+    DNNVersion versionEnum = DNN_VERSION_MAP[nnVersion_];
+
+    // Access the input configuration based on the DNN version
+    std::vector<std::string> input_names;
+    int shapeFeatures;
+
+    auto dnnConfig = dnnInputConfigurations.find(versionEnum);
+    if (dnnConfig != dnnInputConfigurations.end()) {
+        // Valid DNN version found, retrieve input names and shape features
+        input_names = dnnConfig->second.first;
+        shapeFeatures = dnnConfig->second.second;
+    } else {
+        // Invalid DNN version
+        edm::LogError("HGCalTICL_TracksterLinkingByGNN") << "Architecture not defined: " << nnVersion_ ;
+        return;
+    }
+    
+    // Print the retrieved input configuration
+    std::cout << "Input names:";
+    for (const auto& name : input_names) {
+        std::cout << " " << name;
+    }
+    std::cout << std::endl;
+    std::cout << "Shape of features: " << shapeFeatures << std::endl;
+    
+    // Array of data to be filled as a network input. Should be a float array of flattened values.
+    cms::Ort::FloatArrays data;
+    // Network input shapes.
+    std::vector<std::vector<int64_t>> input_shapes;
+    
+
+    auto const &tracksters = input.tracksters;
+    auto const &layerClusters = input.layerClusters;
+    long int N = input.tracksters.size();
+
+    LogDebug("HGCalTICL_TracksterLinkingByGNN") << "Number of tracksters in event: " << N << std::endl;
+
+    if (N < 2) {
+        // do not run the network - return the original tracksters
+        LogDebug("HGCalTICL_TracksterLinkingByGNN") << "Number of tracksters less than 2 - no linking is done." << std::endl;
+        
+        linkedTracksterIdToInputTracksterId.resize(N);
+        std::vector<unsigned int> linkedTracksters;
+
+        for (int trackster_id = 0; trackster_id < N; trackster_id++) {
+            linkedTracksterIdToInputTracksterId[trackster_id].push_back(trackster_id);
+            linkedTracksters.push_back(resultTracksters.size());
+            resultTracksters.push_back(input.tracksters[trackster_id]);
+            outputSuperclusters.push_back(linkedTracksters);
+        }
+        return;
+    }
+
+    std::vector<float> features;
+    std::vector<float> edge_features;
+    
+    std::vector<float> node_degrees;
+    std::vector<float> degree_centr;
```

Comment:
can they be reserved?

Reply:
fixed for features, node_degrees and degree_centr, but not  edge_features (unknown number of edges at this point)